Query Filters (#1)
amercader/ckanext-search: doc/filters.md

# CKAN Query Filters spec

The goal of this specification is to provide a common syntax for defining filters
that need to be applied to queries performed. It is not tied to a particular API
endpoint or search backend, and it aims to be simple and extensible.

## Overview

Filters are always provided as a dictionary or as a list of dictionaries.

When provided as a dictionary, the top level keys can either be:

* A field name
* A top level operator: one of `$or` or `$and`

For example:

```
"filters": {
    "incident": "noise complaint",

>     "$or": {
>         "resolution": ["unresolved", "in progress"],
>         "year": {"gt": 2024}
>     },

Reviewer: ```suggestion
    "$or": [
        {"resolution": ["unresolved", "in progress"]},
        {"year": {"gt": 2024}}
    ],
```
as per discussion below

    "sector": [2, 8, 10, {"gte":15, "lte": 30}, 35],
}
```

When provided as a list of dictionaries, these are combined with the OR operator. For instance:

```
"filters": [
    {"field1": "value1"},
    {"field2": "value2"},
]
```

is interpreted as:

```
"filters": {
  "$or": [
    {"field1": "value1"},
    {"field2": "value2"},
  ]
}
```

## Individual filters

Individual filters have the form:

```
  {
    "<field_name>": {"<operator1>" : <value1>, "<operator2>" : <value2>...}
  }

```

> Note that fields starting with the character `$` must be escaped with another `$` character:
> 
> ```
> filters: {
>     "$$some_field": "some_value"
> }
> ```

Reviewer: ````suggestion
Note that fields starting with the character `$` must be escaped with another `$` character, for example a `$AU` field would be given as:

```
filters: {
    "$$AU": {"lt": 150}
}
```
````

# CKAN Query Filters spec

The goal of this specification is to provide a common syntax for defining filters
that need to be applied to queries performed. It is not tied to a particular API
endpoint or search backend, and it aims to be simple and extensible.

## Overview

Filters are always provided as a dictionary or as a list of dictionaries.

When provided as a dictionary, the top level keys can either be:

* A field name
* A top level operator: one of `$or` or `$and`

For example:

```

> "filters": {
>     "field1": "value",
>     "$or": [
>         {"field2": ["value1", "value2"]},
>         {"field3": {"gte": 9000, "lt": 1000}}
>     ]
> }

Reviewer: If we use real looking examples it could make the docs more engaging:

```suggestion
"filters": {
    "incident": "noise complaint",
    "$or": {
        "resolution": ["unresolved", "in progress"],
        "year": {"gt": 2024}
    },
    "sector": [2, 8, 10, {"gte":15, "lte": 30}, 35],
}
```

I like this as a first example because it covers most of the functionality we'll describe: a short-form "eq" filter operation, a logical filter operation, a short-form "in" field operation, a simple "gt" field operation, and a more complex but understandable combined logical "or" and "in" field operation

```
  {
    "<field_name>": {"<operator1>" : <value1>, "<operator2>" : <value2>...}
  }

```

Note that fields starting with the character `$` must be escaped with another `$` character:

```
filters: {
    "$$some_field": "some_value"
}
```

If the operator is omitted, the `eq` (for equality) operator is assumed:

```
  {
    "field1": "value1"
  }

```

is equivalent to:

```
  {
    "field1": {"eq" : "value1"}
  }

```

If filters contain a list of values, the `in` operator is assumed, e.g.:

```
  {
    "field1": ["value1", "value2", "value3"]
  }

```

is interpreted as:

```
  {
    "field1": {"in": ["value1", "value2", "value3"]}
  }

```

## Built-in operators

The following operators are supported by default:

| Notation | Operation                            |
| -------- | ------------------------------------ |
| `eq`     | Equal                                |
| `gt`     | Greater than                         |
| `gte`    | Greater than or equal                |
| `lt`     | Lower than                           |
| `lte`    | Lower than or equal                  |
| `in`     | Matches one of the provided values   |

TODO: Additional operators can be added by plugins

> ## Top level operators

Author: @wardi consider these filters:

```
"filters": {
    "field1": "value1",
    "$or": [
        {"field2": ["value21", "value22"]},
        {"field3": {"gte": 9000, "lt": 1000}}
    ]
}
```

Would this be how you'd expect them to be parsed?

```
filters = FilterOp(op='$and', field=None, value=[
    FilterOp(op='eq', field='field1', value='value1'),
    FilterOp(op='$or', field='None', value=[
        FilterOp(op='in', field='field2', value=['value21', 'value22']),
        FilterOp(op='$and', field=None, value=[
            FilterOp(op='gte', field='field3', value=9000),
            FilterOp(op='lt', field='field3', value=10000),            
        ])
    ])
])
```

If so perhaps we need to perhaps rename "Top level operators" to "Logial operators" as they can be nested and not just at the top level?

Reviewer: That parsing looks correct.

Naming things is so hard. I suggested "top level" because operators are normally after fields, "logical operators" is good but I can see sites wanting to extend them, so really we need a name for:

- things that can appear as a filter, e.g.
  - `"fieldname": ...`
  - `"$op": ...`
- things that can appear after a field name is given, e.g.
  - `"string"`
  - `42`
  - `true`
  - `["some, "values"]`
  - `{"op": ..., ... }`
  - `[{"op": ..., ... }, "or", "values", ...]`
 
how about **"filter operations"** and **"field operations"**?

- "$op" is a **filter operator** and "op" is a **field operator**
- "filters" values are either a filter operations dict or a list of filter operations dicts

Author: I'm not in love with it, but I don't have any better alternative yet :) 
I'll have another thought but we can go with this for now.

The following default top operators can be used to combine filters:

| Notation | Operation                                  |
| -------- | ------------------------------------------ |
| `$or`    | logical OR, any of the filters can match   |
| `$and`   | logical AND, all of the filters must match |

Top level operators must always contain lists of at least two filters, i.e. all of the 
following are incorrect:

```
{
  "$or": {"field1": "value1"},
}

{
  "$and": [
    {"field1": "value1"}
  ]

> }

Reviewer: We can allow 0- and 1-element `"$or"` and `"$and"` values if we treat them like the python `any()` and `all()` functions.

```python
>>> any([])  # same as "$or": []
False
>>> all([])  # same as "$and": []
True
```
The benefit is when search filters for API calls are generated by code that code doesn't need to special-case the 0 and 1-element cases.

Author: I can see how ignoring empty lists is useful, but I don't understand how considering them False or True works with the actual parsing, ?

For example, what does this parse to?

```
{
   "field1": "value1",
   "$or": []
}
```
Is it something like:
```
field1 = value AND False
```

Reviewer: Yes `false` works. We'd still generate all the FilterOps and the back end would generate something like `where "field1" = 'value1' and false` then return no records for us.

Author: If the goal is make it easier for code generated filters by not chocking on empty lists so users don't have to worry about them, getting no results because you inadvertently sent  an empty list might be more confusing. You effectively need to check that you are not sending empty lists to get some results back, which defeats the purpose of not special casing empty lists right?

Reviewer: Getting nothing back from an empty `$or` list is consistent and shouldn't cause an error.
 - `"color": ["red", "blue"]` might give some results
 - `"color": []` we would expect to give 0 results

Which is the same as:
- `"$or": [{"color": "red"}, {"color": "blue"}]` might give some results
- `"$or": []` we would expect to give 0 results

Getting 0 results is also fine because this might be a smaller part of a larger `$or` condition that we don't want to error out just because one part gives no results.

Reviewer: I like the idea of treating `$and` and `$or` like `all()` and `any()` because I think it's a better interface, but if you think it causes confusion then we could raise an error in the 1-element and 0-element cases instead.

{
  "$and": ["value1", "value2", "value3"]
}

> 

Reviewer: ````suggestion

When a dict is given as the value of an `$or` filter operation each key is treated as a separate part of the OR operation:

```
"filters": {
  "$or": {
    "color": "red",
    "size": 10
  }
}
```

is equivalent to:
```
"filters": {
  "$or": [
    {"color": "red"},
    {"size": 10}
  ]
}
```

This rule does **not** extend to values within a list, e.g.
```
"filters": {
  "$or": [
    {"color": "yellow", "size": 9},
  ]
}
```

is equivalent to:
```
"filters": {
  "color": "yellow", "size": 9
}
```

````

Reviewer: @amercader what do you think about this one? On the one hand it allows users to skip a bunch of extra dicts in their `$or` operations, but on the other hand someone might eventually try to `"$or": {"score": {"gt": 90}, "score": {"lt": 20}}` and our code will **silently ignore** one of the conditions.

Author: I think we need to draw a line somewhere and avoid complicating the spec and the implementations, specially if it can lead to edge cases like the one you mention. I quite like enforcing lists of dicts for the filter operators. Let's test ride this and we can always extend it later.

```

> If filters contain lists of operator/value dictionaries, these are combined with the OR operator, e.g.:

Reviewer: ```suggestion
If a field operation is a lists of values or other field operations, the values are grouped into an "in" field operation and the field operations are ORed together, e.g.:
```

```

>   {
>     "<field_name>": [
>       {"<operator1>" : <value1>}, 
>       {"<operator2>" : <value2>}
>     ]
>   }

Reviewer: ```suggestion
  {
    "year": [1990, 2010, {"gt", 2023}]
  }
```

```

is equivalent to:

```

>   {
>     "$or": [
>       {"field_name": {"<operator1>" : <value1>}}, 
>       {"field_name": {"<operator2>" : <value2>}}
>     ]
>   }

Reviewer: ```suggestion
  {
    "$or": [
      {"year": {"in": [1990, 2010]}}, 
      {"year": {"gt": 2023}}
    ]
  }
```

```

Operators inside the same dict are combined with the AND operator, i.e. the following 

```
  {

>     "<field_name>": {"<operator1>" : <value1>}, {"<operator2>" : <value2>}

Reviewer: ```suggestion
    "last_name": {"gte" : "Jeong", "lte": "Romano"}
```

  }
```

is equivalent to:

```

>     {
>       "$and": [
>         "<field_name>": {"<operator1>" : <value1>},
>         "<field_name>": {"<operator2>" : <value2>}
>       ]
>     }

Reviewer: ```suggestion
    {
      "$and": [
        {"last_name": {"gte" : "Jeong"}},
        {"last_name": {"lte" : "Romano"}}
      ]
    }
```
